Improve vectorization of novelty computation

mberr opened the pull request:
This PR improves the vectorization of novelty computation for `predict_heads` / `predict_tails`.

Fixes #49

Still to do:

- [x] Provide fast implementation for scoring/sorting all possible triples (@mberr)
- [x] Provide in documentation tutorial about making predictions (@cthoyt)

mberr committed:
Add invert_mapping to __all__

mberr committed:
Improve vectorization

mberr committed:
Improve docstring

cthoyt commented on src/pykeen/models/base.py:
```diff
@@ -299,6 +300,39 @@ def predict_scores_all_tails(
             scores = torch.sigmoid(scores)
         return scores
 
+    def _get_novelty_mask(
```
can we actually make this a static function? I realized it might also be useful to add an additional column to the prediction dataframe to show if a given triple appeared in either the testing/validation set. Since these triples factories aren't stored in the model itself, then this logic will have to be applied by an external function to the model itself

mberr replied:
Sure.

There might also be overlap with the LCWA preprocessing code (where we group triples that share the same `(h,r)` / `(r,t)` part).

mberr replied:
Done in 54c720a

cthoyt replied:
Do you want to look into combining the code in the scope of this PR? Or are we done?

mberr replied:
Let's move it to an issue :wink: 

I suppose @lvermue might help us here, since he looked deeply into fast grouping :slightly_smiling_face:

lvermue replied:
I'll address this in #61 .

mberr committed:
Extract function from method

mberr committed:
Add unittest for get_novelty_mask

mberr committed:
Fix line-too-long

cthoyt committed:
Add ability to include testing novelties

cthoyt committed:
Add tests for prediction

cthoyt committed:
Merge branch 'master' into improve-novelty-computation

mberr commented on tests/test_pipeline.py:
```diff
@@ -23,6 +24,22 @@ def test_pipeline(self):
+        nations = Nations()
+        testing_mapped_triples = nations.testing.mapped_triples.to(model.device)
+
+        tails_df = model.predict_tails('brazil', 'intergovorgs', testing=testing_mapped_triples)
```
@cthoyt We can additionally check for the df format (e.g. columns, num_rows). This will help us to detect "API-violating" changes, and some users might rely on the dataframe format for further processing.

cthoyt replied:
Done in b91425b

cthoyt committed:
Add more tests

Also found bug where if you don't want novelties but you have a testing map it added it anyway. Fixed with improvement to conditional

mberr committed:
Add utility to predict high-score triples

mberr committed:
Adjust warning message

cthoyt committed:
Use itertools to combine independent nested loops

cthoyt committed:
Add docs

cthoyt reviewed (changes requested):
The result of `model.predict_top_k_triples(5)` gives back only a tensor. It would be nice to make it possible to also return this as a dataframe, including the labels for entities as well as the scores themselves. Using the same implementation, it should also be possible to make K optional, and return all triples. This might also motivate a name change for this function

```python
>>> model.predict_top_k_triples(5)
tensor([[ 0, 40, 12],
        [ 0, 37, 12],
        [ 0, 27, 12],
        [ 0, 53, 12],
        [ 0, 15, 12]])
```

mberr committed:
Inline variable

mberr committed:
Rename function and make k Optional

mberr committed:
Add warnings

mberr committed:
Allow computing scores for all triples

mberr committed:
Fix typo

mberr commented on src/pykeen/models/base.py:
```diff
@@ -490,6 +556,89 @@ def predict_scores_all_heads(
+        Compute scores for all triples, optionally returning only the k highest scoring.
+
+        .. note ::
+            This operation is very expensive for reasonably-sized knowledge graphs.
```
@cthoyt Please check whether this is the correct way to get this nicely displayed in the doc :wink:

cthoyt replied:
There's a bit of a typo - you need to make sure the `::` comes right after the directive like in:

```
# wrong:
.. note ::

# right:
.. note::
```

If you can manage to fit it all on one line, it can come directly after the `::`, otherwise, there should be a blank like and an ident
```
.. note::

    block text
    goes here
```

cthoyt replied:
I fixed this in 83d6309 and also updated the autogeneration of docs in 8bb8742 so this actually gets turned into a page in the reference

mberr committed:
Fix docstring

mberr commented on src/pykeen/models/base.py:
```diff
@@ -490,6 +556,89 @@ def predict_scores_all_heads(
+            The number of triples to return. Set to None, to keep all.
+
+            .. warning ::
+                This may lead to huge memory requirements.
```
Same here

mberr committed:
Fix docstring

mberr committed:
Change variable name for consistency

mberr committed:
Fix accumulation device when scoring all triples

mberr committed:
Actually fix device

mberr committed:
Improve docstring

cthoyt reviewed (approved):
Idk what's going on with readthedocs but it seems like a non-deterministic PyPI issue, so this is okay to merge. Thanks, Max

mberr committed:
do not track gradients for score_all_triples

mberr commented on docs/source/tutorial/making_predictions.rst:
```diff
@@ -0,0 +1,39 @@
+
+1. :func:`pykeen.models.Model.predict_tails` for a given head/relation pair
+2. :func:`pykeen.models.Model.predict_heads` for a given relation/tail pair
+3. :func:`pykeen.models.Model.predict_top_k_triples` for prioritizing links
```
Now it's called `score_all_triples`

cthoyt committed:
Fix documentation generation

cthoyt committed:
Fix name

cthoyt committed:
Add additional test

@mberr the one with batch_size=1 fails

cthoyt committed:
Reorganize test pipeline

cthoyt committed:
Update test_models.py

cthoyt committed:
Add function for labeling tensor

cthoyt commented on src/pykeen/models/base.py:
```diff
@@ -490,6 +556,133 @@ def predict_scores_all_heads(
+            result = result[ind]
+            # scores = scores[ind]
+
+            return result
```
still need to add a keyword argument to include the scores as the fourth column of this tensor, or return a second 1-d tensor with the scores in order (either are fine for me)

mberr replied:
Done in a2e285c 

I did not use a keyword arguments, since just always returning `(result, scores)` does not need any extra code, and the performance/memory-wise there is not much to gain with having an extra code path based on knowing we do not care about the scores.

cthoyt replied:
do they need to be sorted by ind?

cthoyt commented on src/pykeen/models/base.py:
```diff
@@ -490,6 +556,133 @@ def predict_scores_all_heads(
+
+            return result
+
+    def make_labeled_df(self, tensor: torch.LongTensor) -> pd.DataFrame:
```
@mberr notice I've added the implementation for converting a tensor of indexes to a pandas dataframe including the labels

cthoyt commented on src/pykeen/triples/triples_factory.py:
```diff
@@ -520,6 +521,18 @@ def _word_cloud(self, *, text: List[str], top: int):
         word_cloud = WordCloud()
         return HTML(word_cloud.get_embed_code(text=text, topn=top))
 
+    def tensor_to_df(self, tensor: torch.LongTensor) -> pd.DataFrame:
```
will need an update to handle tensors that are not only 3 columns, as in the case when a score column (or other stuff, like novelties with respect to testing and training set) might be included

cthoyt commented on tests/test_models.py:
```diff
@@ -502,6 +510,49 @@ def _check_constraints(self):
+        # this is only done in one of the models
+        self._test_score_all_triples(k=15, batch_size=16)
+
+    def test_score_all_triples_singleton_batch(self):
```
@mberr I added this test, because using batch size of 1 fails

mberr replied:
should be fixed by 5d6d2d4

cthoyt committed:
Pass flake8

cthoyt reviewed (changes requested):
See my previous comments - I found a scenario where `score_all_triples` fails and added a unit test to reflect it. I think we should also update the `score_all_triples` function to output the scores themselves before merging this PR as well

mberr committed:
Return scores

mberr committed:
Add dictionary inversion utility

mberr committed:
Create properties for ID-to-label conversion

mberr committed:
Accelerate tensor_to_df

mberr committed:
Add unittest

cthoyt commented on src/pykeen/models/base.py:
```diff
@@ -490,6 +556,133 @@ def predict_scores_all_heads(
+            result = result[ind]
+            # scores = scores[ind]
+
+            return result, scores
```
can we have an option `as_df` that would add the labels and combine with the scores into a single pandas dataframe? or should we make a different function for that?

mberr committed:
Fix unittest

mberr committed:
Extend unittest

mberr committed:
Fix unittest

mberr committed:
Fix case where number of elements in batch is smaller than k

mberr committed:
Improved input validation

mberr committed:
Fix line-too-long

cthoyt committed:
Merge branch 'master' into improve-novelty-computation

cthoyt committed:
Fix tuple addition

cthoyt committed:
Make sure scores get sorted too!

@mberr please confirm this change makes sense

cthoyt committed:
Return sweet sweet dataframes

cthoyt committed:
Update docs

cthoyt committed:
Factor out update logic

cthoyt committed:
Update docs

mberr committed:
Update base.py

Simplify code by using that torch.sort returns the sorted array *and* the indices.

cthoyt committed:
Fix dataframe processing

cthoyt committed:
Begin implementing df postprocessing for score_all_triples

- unit tests written
- some of implementation written (abstracted some of the code from the previous implementation for sided predictions)
- I left a todo for @mberr to implement get_novelty_all_mask()

cthoyt committed:
Add predict all to notebook

[skip ci]

cthoyt commented on src/pykeen/models/base.py:
```diff
@@ -71,6 +73,110 @@ def _extend_batch(
+    return np.isin(element=query_ids, test_elements=known_ids, assume_unique=True, invert=True)
+
+
+def get_novelty_all_mask(
```
@mberr please implement this function such that it takes a query set of triples and then returns if each is in mapped_triples

mberr committed:
Add .values

mberr committed:
Fix type annotation

mberr committed:
Add quick and dirty solution for get_novelty_all_mask

cthoyt committed:
Fix returns

cthoyt committed:
Fix tuple checking

cthoyt committed:
Update notebooks

cthoyt committed:
Improve cosmetics of membership check

[skip ci]

cthoyt commented on src/pykeen/models/base.py:
```diff
@@ -490,6 +611,195 @@ def predict_scores_all_heads(
+        :return: Parallel arrays of triples and scores
+        """
+        # initialize buffer on cpu
+        scores = torch.empty(self.num_relations, self.num_entities, self.num_entities, dtype=torch.float32)
```
should we specify the device here?

mberr replied:
From the [documentation](https://pytorch.org/docs/stable/generated/torch.empty.html?highlight=empty#torch.empty)

``` python
torch.empty(*size, out=None, dtype=None, layout=torch.strided, device=None, requires_grad=False, pin_memory=False) → Tensor
```

and 

> `device (torch.device, optional)` – the desired device of returned tensor. Default: if `None`, uses the current device for the default tensor type (see `torch.set_default_tensor_type()`). device will be the CPU for CPU tensor types and the current CUDA device for CUDA tensor types.

So setting it to `cpu` should be the safer option.

cthoyt committed:
Add more tests becuase I have no idea what's wrong

cthoyt committed:
Fix negation
